[JBWS-4080] simplified the algorithm

rsearls opened the pull request.

asoldano commented on modules/server/src/main/java/org/jboss/wsf/stack/cxf/deployment/WSDLFilePublisher.java:
```diff
@@ -118,28 +118,53 @@ public void publishWsdlFiles(QName serviceName, String wsdlLocation, Bus bus, Li
+         do {
+            dirList.add(ancestorDir);
+            ancestorDir = ancestorDir.getParentFile();
+         } while(!wsdlRootDirStr.equals(ancestorDir.getAbsolutePath()));
```
Is it right to compare an absolute path here with the canonical path of the initial wsdlRootDir? Btw, not !wsdlRootDir.equals(ancestorDir) ?

asoldano commented on modules/server/src/main/java/org/jboss/wsf/stack/cxf/deployment/WSDLFilePublisher.java:
```diff
@@ -118,28 +118,53 @@ public void publishWsdlFiles(QName serviceName, String wsdlLocation, Bus bus, Li
+               if (wsdlFileList != null) {
+                  for (int i = 0; i < wsdlFileList.length; i++) {
+                     File f = wsdlFileList[i];
+                     if (f.delete()) {
```
what if the file is a directory and is not empty? IOW, you're assuming here that each directory is nested within a single directory only.

rsearls commented:
> Besides them, did you actually test the changes and were able to verify the fix actually solve the performance issue?
I do not have a Windows machine to test the performance issue.  We don't see an issue with our testsuite.
It would be difficult to determine how the user had their system loaded to reproduce the performance
described.   

The customer noted this was not an issue prior to jbossws-cxf-5.1.5 and they returned to jbossws-cxf-5.1.4.
I reviewed the 5.1.4 code, that code was not using recursion in the deletion process.  It was a very flat processing style.  Based upon that and on the jira description it seemed a reasonable conclusion that the recursion was most likely a factor and could be taken out of the equation by returning to a more flat processing style.

I did run a couple of individual tests and the results looked good.  I run the integration-tests but was looking
for failures not strictly the contents of the wsdl dir.  Unpon re-running the integration tests I did see tests with
more subdirectories.  I've made changes to my code to address that.  I've added a depth-first search for 
direcorties only and return those in the list. 

If you prefer not to take this PR and to investigate the performance side of the issue, I will be happy
to remove this PR.

asoldano commented:
Hi @rsearls ,
ok, the issue with more subdirectories is addressed, thanks.
Now I think I see the point regarding the recursion: CXF' FileUtils.removeDir(File d) internally performs a deep-first visit of the directory to delete the contents. Let's consider a scenario in which we have a EAR with 10 ws JARS in it: Jim's fix in 5.1.5 can cause 10 threads to try deleting the directory structure created for the EAR at the same time using the same procedure. Previously, in 5.1.4 (https://github.com/jbossws/jbossws-cxf/blob/jbossws-cxf-5.1.4.Final/modules/server/src/main/java/org/jboss/wsf/stack/cxf/deployment/WSDLFilePublisher.java#L117), those thread would have delete "their" own JAR dir only.
Now, re-reading JBWS-4080 description, I realize that the actual error in Jim's fix in 5.1.5 was to have each thread for every JAR subdeployment try deleting files for all the subdeployments of the EAR; that is most likely causing contention on the Windows filesystem. As a consequence, I'm not confident your fix here would improve the situation, as again all threads processing subdeployments of the same EAR will be trying to delete the files/dirs for all subdeployments at the same time (and in the same order).
My suggestion would be to implement a mechanism to take track of which subdeployment of the EAR has already been processed and eventually ask for the EAR directory to be deleted only when processing the last WS JAR subdeployment. Alternatively, we could try having each thread delete its own JAR directory and eventually try deleting the EAR directory only if it's empty (meaning all other threads have already run). In any case, we need QA or anybody else with a Windows box to run and validate the new algorithm.

rsearls commented:
We don't currently have a test that has 2 or more modules in an EAR that demonstrates unpublishWsdlFiles being called for each module.  In this PR I would like to add changes to EarSchemaImportTestCase that do 
provide such a testcase.  

The code presented here deletes all files and directories of the EAR in the first pass.  All subsequent calls
to unpublishWsdlFiles will fail the if publishDir.exists() test at line 139 and skip the (unneeded) processing.
I want to move the code lines 160-163 to between 158-159 so that an incorrect log message is not registered.

Are you OK with me making these changes to this PR?

rsearls commented:
Also this PR is not using FileUtil to perform the delete.  Do you want me to make changes to use FileUtil to do that?

asoldano commented:
> We don't currently have a test that has 2 or more modules in an EAR that demonstrates unpublishWsdlFiles being called for each module. In this PR I would like to add changes to EarSchemaImportTestCase that do provide such a testcase.

Sure, feel free to add any test you believe is needed.

> The code presented here deletes all files and directories of the EAR in the first pass. All subsequent calls to unpublishWsdlFiles will fail the if publishDir.exists() test at line 139 and skip the (unneeded) processing.

IMO the problem here is that multiple subdeployment can be processed at the same time in different threads, so the publishDir.exists() check won't help (say, 2 threads start the unpublish process at the same time, they both "see" the ear dir there and both start deleting the contents...)
The problem is really that each thread should deleting only the dirs/files of its "own" jar and only when that's done, check and possibly delete the parent (ear) dir.

> Also this PR is not using FileUtil to perform the delete. Do you want me to make changes to use FileUtil to do that?

Thanks for having pointed this out, I missed it in my last review pass. Yes, you should be using the FileUtil from CXF. The reason is that the filesystem on Windows sometimes does not let you delete files; to cope with this CXF tries waiting a bit and then re-tries.

rsearls commented:
I've changed the algorithm. I believe this meets the thread processing needs you described and does not use recursion for the deletes.  Step 1, only files and subdirectories of publishDir are removed.  Step 2
publishDir is removed if empty (it should be) and  each ancestor dir upto wsdl is checked for being
empty and removed if it is.  This allows proper cleanup when there are multiple modules I believe.   

Please review my use of a synchronized block at 163.  I think its needed here to avoid having one thread remove an ancestor directory while another thread is suspended and in the middle of doing the same.

FileUtils is used for deleting.  I added code to EarSchemaImportTestCase so we have a testcase that has 2 modules to clean up.

asoldano commented:
I'll carefully review on Monday. However, please note in the mean time that each deployment gets its own instance of the WSDLFilePublisher , so a synchronized(this) in there is not going to prevent parallel processing of multiple undeployments in different threads.

rsearls commented:
So if I have an EAR with 2 modules M1 and M2, each of these modules gets
its own instance of WSDLFilePublisher or the EAR gets an instance of
WSDLFilePublisher and M1 and M2 use that
instance.  If its the first of these two scenarios how do you recommend the
code prevent
a failure like the one I described in previous comments?

On Fri, Nov 3, 2017 at 5:57 PM, Alessio Soldano <notifications@github.com>
wrote:

> I'll carefully review on Monday. However, please note in the mean time
> that each deployment gets its own instance of the WSDLFilePublisher , so a
> synchronized(this) in there is not going to prevent parallel processing of
> multiple undeployments in different threads.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jbossws/jbossws-cxf/pull/32#issuecomment-341837228>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAe789Y5m3i2XS4fWJadl5gp9ODSyMGHks5sy4wugaJpZM4QNeHJ>
> .
>

asoldano commented on modules/server/src/main/java/org/jboss/wsf/stack/cxf/deployment/WSDLFilePublisher.java:
```diff
@@ -118,24 +119,61 @@ public void publishWsdlFiles(QName serviceName, String wsdlLocation, Bus bus, Li
+            File parent = publishDir;
+            while(!wsdlRootDirStr.equals(parent.getCanonicalPath())) {
+
+               synchronized(lock) {
```
I'm wondering... shouldn't the synchronized block contain the whole while loop here? IOW, shouldn't the removal of parent and grand-parents' (contents) be considered as an atomic operation?

rsearls commented:
Yes your right Alessio.  The initial assignment of "parent" and the while loop all needs to be protected inside
the sync block.     I've made that code change.  Thanks

rsearls commented:
Please wait on this.  I just found an issue with the testcase.

rsearls commented:
The test is fixed now.  This should be good to go.

asoldano commented:
I've reverted this in https://github.com/jbossws/jbossws-cxf/commit/e2dd1259ff3c59072ccc4839157f17a7b01b79aa . I've been thinking about this for some time and eventually figured out that it would introduce a potential security issue/risk. Basically, considering the files are deleted in a while loop as long as the current parent directory name is not equal to a given directory name based on the deployment name, anything that could make that check fail (e.g because something is changed on the filesystem even by mistake) would result in the procedure there wiping out everything till the root of the filesystem, assuming the virtual machine runs with enough user permissions.
We need a more secure approach here.
